xwiki-platform-core/xwiki-platform-extension/xwiki-platform-extension-index/src/main/java/org/xwiki/extension/index/internal/listener/ExtensionIndexApplicationStartedListener.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.extension.index.internal.listener;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.apache.solr.client.solrj.SolrClient;
+import org.apache.solr.client.solrj.SolrQuery;
+import org.apache.solr.client.solrj.SolrServerException;
+import org.apache.solr.client.solrj.response.QueryResponse;
+import org.apache.solr.common.SolrDocument;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.extension.ExtensionId;
+import org.xwiki.extension.index.internal.ExtensionIndexSolrCoreInitializer;
+import org.xwiki.extension.index.internal.ExtensionIndexSolrUtil;
+import org.xwiki.extension.repository.CoreExtensionRepository;
+import org.xwiki.extension.repository.LocalExtensionRepository;
+import org.xwiki.observation.EventListener;
+import org.xwiki.observation.event.ApplicationStartedEvent;
+import org.xwiki.observation.event.Event;
+import org.xwiki.search.solr.Solr;
+import org.xwiki.search.solr.SolrException;
+import org.xwiki.search.solr.SolrUtils;
+
+/**
+ * Cleanup the extension index of entries related to artifacts that are not available anymore (e.g., upgrade or manual
+ * manipulation of the directories storing jar files).
+ *
+ * @version $Id$
+ * @since 15.9RC1
+ * @since 15.5.3
+ */
+@Component
+@Singleton
+@Named("ExtensionIndexApplicationStartedListener")
+public class ExtensionIndexApplicationStartedListener implements EventListener
+{
+    @Inject
+    private CoreExtensionRepository coreExtensionRepository;
+
+    @Inject
+    private LocalExtensionRepository localExtensionRepository;
```

Comment:
I would check the InstalledExtensionRepository instead.

Reply:
Updated